### PR TITLE
Adopt lintrunner for linting

### PR DESCRIPTION
**Pull request by justinchuby**
### Description

- Adopt lintrunner for linting
- Update CI and readme
- Run `lintrunner -a` on all files
- Explicitly import all proto classes in the onnx namespace to (1) make mypy happy and (2) avoid import *
- Add `.editorconfig`

### Motivation and Context

Lint everything with a single command and config

Now we don't need the custom grep script to check namespaces:

(demo example)

image

### CODE REVIEW

**Comment by p-wysocki** [comment]
Hi! Should I add this PR to ONNX 1.13 milestone? The CF is on 18th November.

**Comment by justinchuby** [comment]
> Hi! Should I add this PR to ONNX 1.13 milestone? The CF is on 18th November.

Thanks, personally I think it’s fine to not include it in the release, as it is more about dev tooling. Unless @jcwchen or others feel differently.

**Review by jcwchen — COMMENTED** [review]
> Thanks, personally I think it’s fine to not include it in the release, as it is more about dev tooling. Unless @jcwchen or others feel differently.

+1 this PR is more related to lint and CI stuff so it is fine to either way (include or not include). I don't have a strong opinion about it.

File: .lintrunner.toml
Comment (jcwchen) [inline .lintrunner.toml:69]: Could you please provide more context regarding this TODO (for instance, what is the blocker that we cannot apply it now)? Shall we create an issue to track it?

File: pyproject.toml
Comment (jcwchen) [inline pyproject.toml:19]: Why turn off this?

File: .lintrunner.toml
Comment (jcwchen) [inline .lintrunner.toml:211]: Previously it was black==22.3?

**Review by jcwchen — COMMENTED** [review]
A few comments. I have a concern that it seems lintrunner is not widely-used yet: https://github.com/suo/lintrunner, although PyTorch is using it. Typically we don't want to introduce too many third-party tools for better management unless it is very useful (saves engineering time) and well-maintained. I hope lintrunner will be kept updated and maintained.

**Comment by justinchuby** [comment]
lintrunner has been stable and used with much success in PyTorch. Since it offers a lot of flexibility in how we manage all of our linters + provides a single command for linting and formatting (`lintrunner`, `lintrunner -a`) + offers result reporting in GitHub, I think it’s adding value in making linting easier and more consistent between local and CI.

File: tools/update_doc.bat
Comment (justinchuby) [reply tools/update_doc.bat:20]: One of the lint rules standardized line endings to POSIX \n

File: .lintrunner.toml
Comment (justinchuby) [reply .lintrunner.toml:211]: I updated to the latest version. Results are unchanged

File: pyproject.toml
Comment (justinchuby) [reply pyproject.toml:19]: I recall there were too many errors. I will double check

File: .lintrunner.toml
Comment (justinchuby) [inline .lintrunner.toml:256]: Use clang-format==13.0.0

File: pyproject.toml
Comment (justinchuby) [reply pyproject.toml]: Turned back on

File: .lintrunner.toml
Comment (justinchuby) [reply .lintrunner.toml]: Done. Added inline and https://github.com/onnx/onnx/issues/4664

File: onnx/numpy_helper.py
Comment (jcwchen) [inline onnx/numpy_helper.py:27]: L9: Do you know why mypy in lintrunner is complaining here (missing attribute e.g., `MYPY/[attr-defined]: Module "onnx" has no attribute "AttributeProto"`)? Does it mean that it needs to build onnx from source as well, just like before?

File: onnx/reference/reference_evaluator.py
Comment (jcwchen) [inline onnx/reference/reference_evaluator.py:295]: I thought the new lintrunner check will be basically same as previous settings, but I do see there are some new errors -- May I understand what are the new checks that bring new error?

**Review by jcwchen — COMMENTED** [review]
Thank you for the quick update. The updates look good to me. Just left around round of reviews.

File: onnx/__init__.py
Comment (justinchuby) [inline onnx/__init__.py:13]: Explicit import here

File: pyproject.toml
Comment (justinchuby) [inline pyproject.toml:38]: Don't check the pbs. they are a headache.

File: pyproject.toml
Comment (justinchuby) [inline pyproject.toml:73]: Just sorted

File: tools/protoc-gen-mypy.py
Comment (justinchuby) [inline tools/protoc-gen-mypy.py:37]: This is needed for the pyi files

File: onnx/numpy_helper.py
Comment (justinchuby) [reply onnx/numpy_helper.py:27]: Just fixed by importing explicitly. It seems to be mypy was not able to follow the import * imports. Not sure what happened previously but this has been a source of mypy errors we haven seen both in this repo and in projects that uses onnx

File: onnx/reference/reference_evaluator.py
Comment (justinchuby) [reply onnx/reference/reference_evaluator.py:295]: It could be how the linter is run, but mypy was not known for being super deterministic from experience. It could be that in this run mypy has got more info about these vars that it can confidently emit an error vs previously treating the type as Any. Just a hypothesis. Most new errors seem legit to me.

**Review by jcwchen — APPROVED** [review]
Overall looks good to me. Please get approval from @onnx/sig-operators-approvers  and ~@onnx/steering-committee~  as well. Thanks!

**Comment by justinchuby** [comment]
@gramalingam PTAL

**Comment by justinchuby** [comment]
The PR should now be ready for your approval @gramalingam

**Review by gramalingam — APPROVED** [review]

**Review by liqunfu — APPROVED** [review]
LGTM

**Pull request merged**
